### PR TITLE
Felix Distribution Mechanism

### PR DESCRIPTION
**Pull request by KPrasch**
Based on #830  .... sounds like a 30-day workout program.

Also updates to Web3 `5.0.0a8`

Here is the new character class definition :
 <https://github.com/nucypher/nucypher/pull/812/files#diff-9f8c085f8e0727aed7eed6c267d99904R119>

*Note*: Still needs documentation; Draft steps to run Felix:

```
nucypher deploy contracts --provider-uri <URI> --poa (if poa network)
```

```
nucypher felix init --provider-uri <URI> --poa (if poa network) --network <DOMAIN> --checksum-address <ADDR OF WALLET WITH TOKENS TO DISTRIBUTE>
nucypher felix createdb
nucypher felix run
```

`POST` to `localhost:6151/register` | `body` `{'address':  YOUR CHECKSUM ADDRESS}`

*Note* - Running with geth may be blocked by gas management. See automated testing for other examples.

*Update*

Opens new Issues: #870 #880 #881 #882 #883 #884 #885 #886 #887 #888 #889

### CODE REVIEW

File: nucypher/blockchain/eth/utils.py [hunk 1]
```diff
@@ -14,33 +14,247 @@
...
+
+    # conversions to smallest denomination
+    __denominations = {'NUWei': 10 ** __decimals,
+                       'NU': 1}
+
+    def __init__(self, value: int, denomination: str):
+
+        # Calculate smallest denomination and store it
+        divisor = self.__denominations[denomination]
+        self.__value = Decimal(value) / divisor
```
Comment (cygnusv) [inline nucypher/blockchain/eth/utils.py]: I think the use of `Decimal` is only justified if, well, we're going to use decimals. However, the type hint in the constructor suggests that the input value is an `int`. Perhaps that should be changed to `Union[int, str, float]`.

Another thing is that the use of `Decimal` can be problematic due to special values ('NaN', 'Infinity', etc.). Mix that with the lack of type checking and we can have awful things like:
```python
>>> Decimal('Infinity') * 10
Decimal('Infinity')
>>> Decimal('NaN') / 100
Decimal('NaN')
>>> Decimal('NaN') == Decimal('NaN')
False
```
I think the NU class requires validating the input with [`Decimal.is_normal()`](https://docs.python.org/3/library/decimal.html#decimal.Decimal.is_normal). There are also some problems with floats and strings that represent floats. I've added some test for edge cases on top of your PR.

File: nucypher/blockchain/eth/utils.py [hunk 2]
```diff
@@ -14,33 +14,247 @@
...
+        return int(self) + other
+
+    def __add__(self, other) -> 'NU':
+        return NU(int(self) + int(other), 'NUWei')
+
+    def __sub__(self, other) -> 'NU':
+        return NU(int(self) - int(other), 'NUWei')
+
+    def __mul__(self, other) -> 'NU':
+        return NU(int(self) * int(other), 'NUWei')
```
Comment (cygnusv) [inline nucypher/blockchain/eth/utils.py]: What about `__rmul__`?

File: nucypher/blockchain/eth/utils.py (hunk 1 above)
Comment (KPrasch) [reply nucypher/blockchain/eth/utils.py]: Excellent suggestions, I honestly do not know why it escaped me to consider floats :joy:

Comment (KPrasch) [reply nucypher/blockchain/eth/utils.py]: (Moving to PR #830 )

Comment (KPrasch) [reply nucypher/blockchain/eth/utils.py]: :heavy_check_mark:

File: nucypher/characters/chaotic.py [hunk 3]
```diff
@@ -95,3 +106,285 @@ def status():
...
         if not dry_run:
             deployer.run()
+
+
+class Felix(Character):
+    """
+    A NuCypher ERC20 faucet / Airdrop scheduler.
+
+    Felix is a web application that gives NuCypher *testnet* tokens to registered addresses
+    with a scheduled reduction of disbursement amounts, and a RESTful interface
```
Comment (KPrasch) [inline nucypher/characters/chaotic.py]:
``` suggestion
    with a scheduled reduction of disbursement amounts, and an HTTP endpoint
```

File: nucypher/characters/banners.py [hunk 4]
```diff
@@ -1,17 +1,27 @@
...
-    version {}
+_____   __      _________               ______              
+___  | / /___  ___  ____/____  ____________  /______________
+__   |/ /_  / / /  /    __  / / /__  __ \_  __ \  _ \_  ___/
+_  /|  / / /_/ // /___  _  /_/ /__  /_/ /  / / /  __/  /    
+/_/ |_/  \__,_/ \____/  _\__, / _  .___//_/ /_/\___//_/     
+                        /____/  /_/                         
 
+version {}
 """.format(nucypher.__version__)
```
Comment (KPrasch) [inline nucypher/characters/banners.py]: :racing_car:

**Comment by michwill** [comment]
Idk why, but when I hear "Felix", the guy I first think of is https://en.wikipedia.org/wiki/Felix_Dzerzhinsky

File: nucypher/characters/banners.py [hunk 5]
```diff
@@ -1,17 +1,27 @@
 import nucypher
 
+
+NU_BANNER = r'''
```
Comment (michwill) [inline nucypher/characters/banners.py]: Holy shit! Awesome

File: nucypher/blockchain/eth/utils.py (hunk 2 above)
Comment (KPrasch) [reply nucypher/blockchain/eth/utils.py]: :heavy_check_mark:

File: nucypher/blockchain/eth/utils.py [hunk 6]
```diff
@@ -14,33 +14,247 @@
...
+    Token quantity is stored internally as a Decimal in the smallest denomination,
+    and all arithmetic operations use this value.
+    """
+
+    __symbol = 'NU'
+    __decimals = 18
+    __agent_class = NucypherTokenAgent
+
+    # conversions to smallest denomination
+    __denominations = {'NUWei': 10 ** __decimals,
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/blockchain/eth/utils.py]: :heavy_check_mark:

**Comment by KPrasch** [comment]
Going back on WIP - after a few weeks of being open, and several rebases, there have been some breaking changes.

File: nucypher/config/characters.py [hunk 7]
```diff
@@ -127,3 +128,49 @@ def write_keyring(self, password: str, **generation_kwargs) -> NucypherKeyring:
...
+         rest_host=self.rest_host,
+         rest_port=self.rest_port,
+         db_filepath=self.db_filepath,
+        )
+        return {**super().static_payload, **payload}
+
+    def write_keyring(self, password: str, **generation_kwargs) -> NucypherKeyring:
+
+        return super().write_keyring(password=password,
+                                     encrypting=True,
```
Comment (cygnusv) [inline nucypher/config/characters.py]: why `encrypting=True`?

Comment (KPrasch) [reply nucypher/config/characters.py]: Hmm, yeah no reason actually. I just wanted signing power only, although on second though- I probably also want `TLSHostingPower`, too

Comment (cygnusv) [inline nucypher/config/characters.py]: This is related to #668. Currently, `encrypting` creates both signing and encrypting keys, and there's no way to ask for only signing.
```suggestion
                                     encrypting=True,  # TODO: #668
```

**Comment by KPrasch** [comment]
@tuxxy - Following up - <https://github.com/nucypher/nucypher/issues/885>

**Comment by KPrasch** [comment]
Rebased over 4adde682fe58a0c79a6ce31c4517638d14e4d27d

File: nucypher/characters/chaotic.py [hunk 8]
```diff
@@ -95,3 +114,355 @@ def status():
...
+                    session.add(recipient)
+                    session.commit()
+
+            except Exception as e:
+                # Pass along exceptions to the logger
+                self.log.critical(str(e))
+                raise
+
+            else:
+                return Response(status=200)  # TODO
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/chaotic.py]: Issue #887

File: nucypher/characters/chaotic.py [hunk 9]
```diff
@@ -95,3 +114,355 @@ def status():
...
+                self.log.info("No eligible recipients this round.")
+                return
+
+        # Discard invalid addresses, in-depth
+        invalid_addresses = list()
+
+        def siphon_invalid_entries(candidate):
+            address_is_valid = eth_utils.is_checksum_address(candidate.address)
+            if address_is_valid:
+                invalid_addresses.append(candidate.address)
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/chaotic.py]: Sounds like you understanding is correct.  There is a missing `not` inside this helper function.  

I'm using the `siphon_invalid_entries` to return `True` is a given element is a valid checksum address.  Used as the filter function, it will remove invalid addresses, and track the invalid ones in a list, resulting in `candidates`.

[the remark this answers is not in the document]
Comment (jMyles) [reply nucypher/characters/chaotic.py]: Sure - but does that mean that `invalid_addresses` isn't doing anything?

[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/chaotic.py]: Yes.  The resulting list `invalid_addresses` is  pretty much just used to generate a log message.

[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/chaotic.py]: Do you think we want to prune the database of bad entries here, or what are you thinking?

File: nucypher/cli/characters/felix.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+                                                       db_filepath=db_filepath,
+                                                       domains={network} if network else None,
+                                                       checksum_public_address=checksum_address,
+                                                       no_registry=no_registry,
+                                                       registry_filepath=registry_filepath,
+                                                       provider_uri=provider_uri,
+                                                       poa=poa)
+
+        # Paint Help
+        painting.paint_new_installation_help(new_configuration=new_felix_config,
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/characters/felix.py]: :heavy_check_mark:

File: nucypher/cli/characters/felix.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+        return  # <-- do not remove (conditional flow control)
+
+    #
+    # Authenticated Configurations
+    #
+
+    # Domains -> bytes | or default
+    domains = [bytes(network, encoding='utf-8')] if network else None
+
+    # Load Ursula from Configuration File with overrides
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/characters/felix.py]: :heavy_check_mark:

File: nucypher/characters/chaotic.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/chaotic.py]: @jMyles - Here's a stub to follow-up: <https://github.com/nucypher/nucypher/pull/812/commits/d71d76beb35e94790fb5ccfb244051d41b128a62#diff-9f8c085f8e0727aed7eed6c267d99904R410>

File: nucypher/cli/painting.py [hunk 12]
```diff
@@ -43,14 +43,25 @@ def paint_new_installation_help(new_configuration, federated_only: bool = False,
...
+    # Ursula
+    elif character_name == 'ursula' and not federated_only:
         suggested_staking_command = f'nucypher ursula stake'
         how_to_stake_message = f"\nTo initialize a NU stake, run '{suggested_staking_command}' or"
         emitter(message=how_to_stake_message, color='green')
 
-    # Give the use a suggestion as to what to do next...
+    # Everyone: Give the use a suggestion as to what to do next
+    vowles = ('a', 'e', 'i', 'o', 'u')
+    character_name_starts_with_vowel = character_name[0].lower() in vowles
```
Comment (jMyles) [inline nucypher/cli/painting.py]: haha, really went for it

Comment (jMyles) [reply nucypher/cli/painting.py]: Important to know if the Character names starts with a

https://upload.wikimedia.org/wikipedia/commons/6/60/Bank_vole.jpg

File: tests/cli/test_felix.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    TEMPORARY_DOMAIN,
+    TEST_PROVIDER_URI,
+    INSECURE_DEVELOPMENT_PASSWORD,
+    MOCK_CUSTOM_INSTALLATION_PATH_2
+)
+
+
+@pytest_twisted.inlineCallbacks
+def test_run_felix(click_runner, testerchain, federated_ursulas, mock_primary_registry_filepath):
+
```
Comment (jMyles) [inline tests/cli/test_felix.py]: Jeez this is a good test.  Very nice engineering.  I particularly like the clock management and synchronous re-entry to cause the airdrop.  See my comment below about my concerns that this is the only test.

File: nucypher/characters/chaotic.py [hunk 14]
```diff
@@ -95,3 +114,363 @@ def status():
 
         if not dry_run:
             deployer.run()
+
+
+class Felix(Character, NucypherTokenActor):
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/chaotic.py]: I'm 50/50. Unsure about both the name and location.

File: tests/cli/test_felix.py [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+               'FLASK_DEBUG': '1'}
+
+    # Deploy contracts
+    deploy_args = ('contracts',
+                   '--registry-outfile', mock_primary_registry_filepath,
+                   '--provider-uri', TEST_PROVIDER_URI,
+                   '--poa')
+
+    user_input = 'Y\n'+f'{INSECURE_DEVELOPMENT_PASSWORD}\n'*6  # TODO: Use Env Vars
+    result = click_runner.invoke(deploy.deploy, deploy_args, input=user_input, catch_exceptions=False, env=envvars)
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply tests/cli/test_felix.py]: I agree.  We need to find a way to effectively mock contract deployment and execution, which can be applied project wide for massive time saving an accurate increases.

In this case, we can deploy the ERC20, or mock the blockchain altogether here or in a sister-test.

[the remark this answers is not in the document]
Comment (KPrasch) [reply tests/cli/test_felix.py]: Opened: #888

File: nucypher/characters/chaotic.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/chaotic.py]: Opened: #889
